curvefs/src/mds/fs_manager.cpp

```diff
@@ -426,22 +499,57 @@ FSStatusCode FsManager::CreateFs(const ::curvefs::mds::CreateFsRequest* request,
-        //                << ", then delete root inode fail"
-        //                << ", fsName = " << fsName
-        //                << ", ret = " << FSStatusCode_Name(ret);
-        //     return ret;
-        // }
-
-        // ret = fsStorage_->Delete(fsName);
-        // if (ret != FSStatusCode::OK) {
-        //     LOG(ERROR) << "CreateFs fail, insert root inode fail, "
-        //                << "then delete fs fail, fsName = " << fsName
-        //                << ", errCode = " << FSStatusCode_Name(ret);
-        // }
+        // delete recycle dentry and recyle inode if created
+        if (request->has_recycletimehour() && request->recycletimehour() != 0) {
+            FSStatusCode ret2 = metaserverClient_->DeleteDentry(
+                wrapper.GetFsId(), partition.poolid(), partition.copysetid(),
+                partition.partitionid(), ROOTINODEID, RECYCLENAME, addrs);
+            if (ret2 != FSStatusCode::OK) {
+                LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                           << ", then delete recycle dentry fail"
+                           << ", fsName = " << fsName
+                           << ", ret = " << FSStatusCode_Name(ret2);
+                return ret2;
+            }
+
+            FSStatusCode ret3 = metaserverClient_->DeleteInode(
+                wrapper.GetFsId(), RECYCLEINODEID);
+            if (ret3 != FSStatusCode::OK) {
+                LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                           << ", then delete recycle inode fail"
+                           << ", fsName = " << fsName
+                           << ", ret = " << FSStatusCode_Name(ret3);
+                return ret3;
+            }
+        }
+
+        FSStatusCode ret4 =
+            metaserverClient_->DeleteInode(wrapper.GetFsId(), GetRootId());
+        if (ret4 != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                       << ", then delete root inode fail"
+                       << ", fsName = " << fsName
+                       << ", ret = " << FSStatusCode_Name(ret4);
+            return ret4;
+        }
+
+        if (TopoStatusCode::TOPO_OK !=
+            topoManager_->DeletePartition(partition.partitionid())) {
+            LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                       << ", then delete partition fail"
+                       << ", fsName = " << fsName
+                       << ", ret = " << FSStatusCode_Name(ret);
+            return FSStatusCode::DELETE_PARTITION_ERROR;
+        }
+
+        FSStatusCode ret5 = fsStorage_->Delete(fsName);
+        if (ret5 != FSStatusCode::OK) {
+            LOG(ERROR) << "CreateFs fail, update fs to inited fail"
+                       << ", then delete fs fail"
+                       << ", fsName = " << fsName
+                       << ", ret = " << FSStatusCode_Name(ret5);
+            return ret5;
```

Comment:
The clean code here is too scattered and verbose, you can abstract a function which maybe named `CleanResourceAfterCreateFsFailed` and clean related resources in different fail stage.

 `FSStatusCode CleanResourceAfterCreateFsFailed (FSStatusCode status) {
  switch (status) {
  case step1:
  ...
  case step2:
  ...
  }
}
`

Reply:
Good suggestion! Let me rearrange it.